### PR TITLE
[4.2] Switch div to img and use the loading attr. [Media Manager images list]

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #36533 [Solves 1/3 reported problems, the thumbs and the streaming/paginated response are still valid... ].

### Summary of Changes
- Add and intersectionObserver so the browser won't try to fetch every image in the list view (only the ones that are visible)

### Testing Instructions
- Apply this PR and run `npm ci` or d/l the package from github
- navigate to different subfolders in the `images` folder
- Open the browser inspector, select the Network tab and then the img tab. As you scroll in the list view more images will appear both in the list view and in the network/img tab. The Network/img should have less images that the total amount of images in the list (assuming the viewport is not revealing all the images and you haven't scrolled to the end of the list)

### Actual result BEFORE applying this Pull Request

All images fetched immediately

### Expected result AFTER applying this Pull Request

Images are fetched as they enter the viewport

### Documentation Changes Required

No

Ping @laoneo @Fedik @crystalenka

### CODE REVIEW

**Comment by laoneo** [comment]
Should we not just add the `loading="lazy"` attribute to the image tag as the dimensions are available for the images. We should use the platform as much as possible said once a Joomla JS developer (can't remember his name :smile: ).

**Comment by crystalenka** [comment]
> Should we not just add the `loading="lazy"` attribute to the image tag as the dimensions are available for the images. We should use the platform as much as possible said once a Joomla JS developer (can't remember his name :smile: ).

There are no img tags in the media manager currently - all images are loaded as background images of div elements.

**Comment by laoneo** [comment]
What is the benefit of loading them as background images?

**Comment by dgrammatiko** [comment]
> We should use the platform as much as possible said once a Joomla JS developer (can't remember his name 😄 ).

This sounds obnoxious. Why would you ever want to use the browser's provided APIS instead of building them again and having a gazillion of bugs?

Joke apart the `loading=lazy` is an attribute for the img element and as @crystalenka already mentioned the design is asking for a style="background-image: ..."

> What is the benefit of loading them as background images?

No alt (obviously this is an a11y problem) and it was easier to style/append the edit buttons/checkbox on top (?)

**Comment by crystalenka** [comment]
I have absolutely no idea, it seems inefficient to me but I'm not sure. Just commenting why lazy loading wouldn't work as the media manager works now.

**Comment by crystalenka** [comment]
> No alt (obviously this is an a11y problem) and it was easier to style/append the edit buttons/checkbox on top (?)

How wonderful it would be to manage alt text globally! But this is beyond the scope of this issue/pr, haha. 

I am happy to help style the edit buttons over the top of inline images if that is more efficient/a necessary change.

**Comment by dgrammatiko** [comment]
Ok, switched to the image tag and used the loading attribute. The styling is a bit of but some `object-fit: contain;` or something similar should get us back to the previous state (design-wise). @crystalenka @laoneo 

Screenshot 2022-01-03 at 18 30 34

File: administrator/components/com_media/resources/scripts/components/browser/items/image.vue [hunk 1]
```diff
@@ -75,5 +75,21 @@ export default {
...
+        : `${this.item.thumb_path}`;
+    },
+    width() {
+      return this.item.width;
+    },
+    height() {
+      return this.item.height;
+    },
+    altTag() {
+      return `Image filename: ${this.item.name}`;
```
Comment (dgrammatiko) [inline administrator/components/com_media/resources/scripts/components/browser/items/image.vue]: @brianteeman any ideas what would be a good alt tag for the images view in the MM?

**Comment by crystalenka** [comment]
I'll test this and check styling later today or tomorrow. Thank you!!

File: administrator/components/com_media/resources/scripts/components/browser/items/image.vue (hunk 1 above)
Comment (brianteeman) [reply administrator/components/com_media/resources/scripts/components/browser/items/image.vue]: ask the accessibility team

Comment (brianteeman) [reply administrator/components/com_media/resources/scripts/components/browser/items/image.vue]: but its currently completely inaccessible so i doubt it matters

File: administrator/components/com_media/resources/scripts/components/browser/items/image.vue [hunk 2]
```diff
@@ -6,10 +6,7 @@
   >
     <div class="media-browser-item-preview">
       <div class="image-background">
-        <div
-          class="image-cropped"
-          :style="{ backgroundImage: getHashedURL }"
-        />
+        <img class="image-cropped" :src="getURL" :alt="altTag" loading="lazy" :width="width" :height="height" >
```
Comment (Quy) [inline administrator/components/com_media/resources/scripts/components/browser/items/image.vue]:
```suggestion
        <img class="image-cropped" :src="getURL" :alt="altTag" loading="lazy" :width="width" :height="height">
```

**Comment by laoneo** [comment]
I first test shows the attributes loading, width and height in the image. But they are loaded initially in the network tab on FF and chrome.

Additionally has the image  container not the same dimensions as the folders:
image

**Comment by dgrammatiko** [comment]
> Additionally has the image container not the same dimensions as the folders:

Yes, switching to the image element needs some scss fixes

> I first test shows the attributes loading, width and height in the image. But they are loaded initially in the network tab on FF and chrome.

If you roll to https://github.com/joomla/joomla-cms/pull/36550/commits/87d01f8cfcc2cdf8ba64b640b7bcbb4da9a4ffff you'll see that the intersectionObserver is a better fit here. The reason is that there's something weird with the zoom-in/out functionality that triggers the browser to always consider more images in the viewport than what's actually there. Didn't spend any time figuring out what's triggering this (container or item size, or anything else)

**Comment by laoneo** [comment]
> The reason is that there's something weird with the zoom-in/out functionality that triggers the browser to always consider more images in the viewport than what's actually there. Didn't spend any time figuring out what's triggering this (container or item size, or anything else)

Pretty sure you can fix this :wink:

**Comment by crystalenka** [comment]
I have tested this item :white_check_mark: successfully on d903326dfea192046d2028f942fe67ddc0726f17<br /><br />If you add the following CSS rules to .image-cropped it fixes the styling issue in modern browsers:
```
aspect-ratio: 1/1;
object-fit: contain;
```

Aspect ratio is supported in most recent releases of most modern browsers (https://caniuse.com/?search=aspect-ratio), I'm not sure what J4 is supposed to support but if necessary I can do this a different way. This is just the most future-proofed way to do it.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36550">issues.joomla.org/tracker/joomla-cms/36550</a>.</sub>

File: administrator/components/com_media/resources/scripts/components/browser/items/image.vue [hunk 3]
```diff
@@ -6,10 +6,14 @@
   >
     <div class="media-browser-item-preview">
       <div class="image-background">
-        <div
+        <img
           class="image-cropped"
-          :style="{ backgroundImage: getHashedURL }"
-        />
+          :src="getURL"
+          :alt="altTag"
```
Comment (Quy) [inline administrator/components/com_media/resources/scripts/components/browser/items/image.vue:13]: Please consider adding `title` attribute as reported in #36302.

Comment (brianteeman) [reply administrator/components/com_media/resources/scripts/components/browser/items/image.vue:13]: adding alt here has no impact as you can not focus on the issue and adding title here will have no impact and will not resolve that issue for the same reason

**Comment by Quy** [comment]
SVG not displaying correctly.

36550-svg

``

**Comment by infograf768** [comment]
@Quy 
Same result as you in Firefox and Chrome
Screenshot 2022-01-12 at 10 53 03

Same error but different display in Safari:

Screenshot 2022-01-12 at 10 51 05

**Comment by dgrammatiko** [comment]
@laoneo will take over this one

**Comment by laoneo** [comment]
Please test #36930 which is the followup on this one.

**Pull request closed without merging**